enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,213 @@
+
+3. **External OAuth identity providers used as challengers**
+
+External oauth identity providers used as challengers use a [central location](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L118) where login and login failures audit events will be emitted.
+
+4. **Request Header identity provider**
+
+For this case an established trust to the external login proxy exists via means of client certificate trust. This implies that all authentication requests, naturally, have already passed successful authentication in the external 3rd party proxy.
+
+Thus, successful Login events can easily be constructed by creating audit events in the [central request authentication method](https://github.com/openshift/oauth-server/blob/4e63f0f350f28171edc4ae2553c7b852e65efbc6/pkg/authenticator/request/headerrequest/requestheader.go#L34).
+
+Login failure events, however never cause a request header authentication request to be generated in the 3rd party proxy, thus those events cannot be detected in oauth-server for this use case.
+
+### Logout events
+
+Logout events are not visible to oauth-server as the infrastructure in OpenShift does not have a central logout handler mechanism. Instead we have to rely on audit events against deletion of `oauthaccesstoken` resources.
+
+Currently, only a `Status` object is created in an deletion event entry of the following form:
+
+```json
+{
+  "kind": "Event",
+  "apiVersion": "audit.k8s.io/v1",
+  "level": "RequestResponse",
+...
+  "verb": "delete",
+...
+  "responseObject": {
+    "kind": "Status",
+    "apiVersion": "v1",
+    "metadata": {},
+    "status": "Success",
+    "details": {
+      "name": "sha256~sometoken",
+      "group": "oauth.openshift.io",
+      "kind": "oauthaccesstokens",
+      "uid": "444845f0-1348-4ab1-95be-d0644a6bd11c"
+    }
+  },
+...
+}
+```
+
+To log the complete object we have to set `ReturnDeletedObject: true` on the oauth-apiserver REST Store: https://github.com/openshift/oauth-apiserver/blob/6e0f92194d5a25728c826fefb2d99c5e88ebb5e5/pkg/oauth/apiserver/registry/oauthaccesstoken/etcd/etcd.go#L30-L46.
+
+This will allow to introspect the deleted user:
+
+```json
+ {
+    "kind": "Event",
+    "apiVersion": "audit.k8s.io/v1",
+    "level": "RequestResponse",
+...
+	"verb": "delete",
+...
+    "responseObject": {
+      "kind": "OAuthAccessToken",
+      "apiVersion": "oauth.openshift.io/v1",
+      "metadata": {
+...
+		"userName": "exampleuser",
+	  },
+      "userUID": "123",
+      "authorizeToken": "sha256~sometoken"
+    },
+...
+  }
+```
+
+### Event construction
+
+In all cases above a [Info](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/authentication/user/user.go#L20) entity is available with user information which can be used to fill audit event data.
```

Comment:
Can you provide an example of what such an event should look like, then?

Reply:
How can I find that out? `must-gather` and check some similar audit logs and exchange values for what we are expecting?

Reply:
@ibihim yes